### PR TITLE
[mono] MarshalingPInvokeScanner tolerates DLLs without metadata

### PR DESCRIPTION
**Pull request by jandupej**
Addresses https://github.com/dotnet/runtime/issues/89429 where `MarshalingPInvokeScanner` failed when trying to acquire a `MetadataReader` over a DLL that does not have metadata. The failure is shown [here](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-89296-merge-1bbcaf3224704de6b0/tracing_eventpipe/1/console.b87213e1.log?helixlogtype=result).

cc: @kotlarmilos

### CODE REVIEW

File: src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs [hunk 1]
```diff
@@ -106,7 +106,17 @@ private bool IsAssemblyIncompatible(string assyPath, MinimalMarshalingTypeCompat
...
-            MetadataReader mdtReader = peReader.GetMetadataReader();
+            MetadataReader mdtReader;
+
+            try
+            {
+                mdtReader = peReader.GetMetadataReader();
+            }
+            catch(InvalidOperationException)
+            {
+                // If the DLL does not have metadata, the previous call will fail with InvalidOperationException.
```
Comment (kotlarmilos) [inline src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs:117]: Logging the exception may be useful.

**Review by kotlarmilos — APPROVED** [review]
LGTM!

File: src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs (hunk 1 above)
Comment (jandupej) [reply src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs:117]: Agreed.

File: src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs [hunk 2]
```diff
@@ -106,7 +106,18 @@ private bool IsAssemblyIncompatible(string assyPath, MinimalMarshalingTypeCompat
...
+            try
+            {
+                mdtReader = peReader.GetMetadataReader();
+            }
+            catch(InvalidOperationException ex)
+            {
+                // If the DLL does not have metadata, the previous call will fail with InvalidOperationException.
+                Log.LogMessage(MessageImportance.Low, string.Format("Cannot read metadata in file {0}: {1}.", assyPath, ex.Message));
+                return false;
+            }
```
Comment (teo-tsirpanis) [inline src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs:120]:
```suggestion
            if (!peReader.HasMetadata)
            {
                return false;
            }
            MetadataReader mdtReader = peReader.GetMetadataReader();
```

**Comment by kotlarmilos** [comment]
@jandupej I think it is still happening. Check out https://dev.azure.com/dnceng-public/public/_build/results?buildId=352503&view=logs&jobId=da83f4c3-af55-5024-34f6-3ac8d11ac15f&j=da83f4c3-af55-5024-34f6-3ac8d11ac15f&t=28977ecb-8381-5b63-4615-92c3dead4a9b in https://github.com/dotnet/runtime/pull/89428.

**Pull request merged**